internals/daemon/access.go

```diff
@@ -16,40 +16,55 @@ package daemon
 
 import (
 	"net/http"
-	"os"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+const (
+	accessDenied = "access denied"
 )
 
 // AccessChecker checks whether a particular request is allowed.
 type AccessChecker interface {
-	// Check if access should be granted or denied. In case of granting access,
-	// return nil. In case access is denied, return a non-nil error response,
-	// such as Unauthorized("access denied").
-	CheckAccess(d *Daemon, r *http.Request, ucred *Ucrednet, user *UserState) Response
+	// CheckAccess reports whether access should be granted or denied. If
+	// access is granted, return nil. If access is denied, return a non-nil
+	// error such as Unauthorized("access denied").
+	CheckAccess(d *Daemon, r *http.Request, user *UserState) Response
 }
 
-// OpenAccess allows all requests, including non-local sockets (e.g. TCP)
+// OpenAccess allows all requests, including non-local sockets (for example, TCP).
 type OpenAccess struct{}
 
-func (ac OpenAccess) CheckAccess(d *Daemon, r *http.Request, ucred *Ucrednet, user *UserState) Response {
+func (ac OpenAccess) CheckAccess(d *Daemon, r *http.Request, user *UserState) Response {
 	return nil
 }
 
-// AdminAccess allows requests over the UNIX domain socket from the root uid and the current user's uid
+// AdminAccess allows requests over the unix domain socket from the root UID
+// and the current user's UID.
 type AdminAccess struct{}
 
-func (ac AdminAccess) CheckAccess(d *Daemon, r *http.Request, ucred *Ucrednet, user *UserState) Response {
-	if ucred != nil && (ucred.Uid == 0 || ucred.Uid == uint32(os.Getuid())) {
+func (ac AdminAccess) CheckAccess(d *Daemon, r *http.Request, user *UserState) Response {
+	if user == nil {
+		return Unauthorized(accessDenied)
+	}
+	if user.Access == state.AdminAccess {
 		return nil
 	}
-	return Unauthorized("access denied")
+	// An identity explicitly set to "access: read" or "access: untrusted" isn't allowed.
+	return Unauthorized(accessDenied)
 }
 
 // UserAccess allows requests over the UNIX domain socket from any local user
 type UserAccess struct{}
 
-func (ac UserAccess) CheckAccess(d *Daemon, r *http.Request, ucred *Ucrednet, user *UserState) Response {
-	if ucred == nil {
-		return Unauthorized("access denied")
+func (ac UserAccess) CheckAccess(d *Daemon, r *http.Request, user *UserState) Response {
+	if user == nil {
+		return Unauthorized(accessDenied)
 	}
-	return nil
+	switch user.Access {
+	case state.ReadAccess, state.AdminAccess:
```

Comment:
I see that only a ```state.ReadAccess``` exist. At the same time I can see that notices have:

```
    Path:        "/v1/notices",
    ReadAccess:  UserAccess{},
    WriteAccess: UserAccess{}, // any user is allowed to add a notice with their own uid
    GET:         v1GetNotices,
    POST:        v1PostNotices,
```

and ```UserAccess``` validation pass with only ```ReadAccess```. Should the state identify access type not simply say ```UserAccess```. It feels that the read/write control is already available at the API endpoint level ?

Reply:
Yeah, it's a little weird that the AccessChecker is called UserAccess but the level is called "read". The latter was spec'd so I'm going to leave that one. I think I'll leave this as is for now, but in a future PR we could consider renaming UserAccess to ReadAccess.